Adds a topic guide of deprecation policy and versioning (#7731)
sunpy/sunpy: docs/topic_guide/deprecation.rst

> .. _sunpy-topic-guide-deprecation-policy:

Reviewer: ```suggestion
.. _sunpy-topic-guide-deprecation-versioning:
```

> ********************************************
> Deprecation Versioning and Release Practices
> ********************************************

Reviewer: ```suggestion
**********************************************
Deprecations, Versioning and Release Practices
**********************************************
```

Versioning Policy
=================

> Sunpy uses a date-based versioning system with three components, each serving a specific purpose:

Reviewer: date-based?

Author: I took it from here https://github.com/PlasmaPy/PlasmaPy/issues/2726#issue-2365209146, should I remove it?

Reviewer: Well does sunpy follow a date-based system?

Author: Changed it to Semantic.

- **X (LTS Version Number)**: Incremented with every Long Term Support (LTS) release.
- **Y (Release Counter)**: Starts at 0 for LTS releases and increases for intermediate releases.

> - **Z (Bug Fix Number)**: Incremented for bug fix releases.

Reviewer: ```suggestion
- **Z (Bug Fix Counter)**: Incremented for bug fix releases.
```

Here's an example release sequence to illustrate:

- **1.0.0**: An LTS release (November 2024).
- **1.0.1**: A bug fix for the LTS release (December 2024).
- **1.1.0**: A short-support release (May 2025).
- **1.0.2**: A bug fix for the LTS release (June 2025).
- **1.1.1**: A bug fix for the short-support release (June 2025).

> - **1.2.0**: An unplanned short-support release (August 2025).

Reviewer: ```suggestion
```
We have never done this as far as I can tell.

- **2.0.0**: An LTS release (November 2025).

> For detailed information, refer to `SunPy Enhancement Proposal (SEP) 0009 <https://github.com/sunpy/sunpy-SEP/blob/main/SEP-0009.md#deprecations-and-documentation>`_.

Reviewer: ```suggestion
For detailed information, refer to `SunPy Enhancement Proposal (SEP) 0009 <https://github.com/sunpy/sunpy-SEP/blob/main/SEP-0009.md#deprecations-and-documentation>`__.
```

Release Practices
=================

> There will be two planned releases of the SunPy core package per year with at least 4 months (preferably 6 months) between these releases.

Reviewer: ```suggestion
There will be two planned major releases of ``sunpy`` per year with at around 6 months between these releases.
```

> Users can expect these releases occur in the months of May and November based on the current release schedule of the astropy core package

Reviewer: ```suggestion
Users can expect these releases occur in the months of June and December based on the current release schedule of the ``astropy``.
```

> SunPy provides two types of releases:

Reviewer: ```suggestion
``sunpy`` provides two types of releases:
```

> - **Long Term Support (LTS) Releases**: These are released at the start of each year and supported for 12 months or until the next LTS release.

Reviewer: ```suggestion
- **Long Term Support (LTS) Releases**: These are first major release of each year and supported for 12 months or until the next LTS release.
```

- **Short-Support (Non-LTS) Releases**: These are the second release of the year, supported for 6 months or until the next release.

Support periods may be extended beyond these requirements if necessary.

Python and Package Support Policies

> ===================================

Reviewer: This is missing the stuff from https://docs.sunpy.org/en/latest/dev_guide/contents/dependencies.html

Probably need to remove it from that page as well.

Author: > Probably need to remove it from that page as well.

Anything else to be removed?

Reviewer: Can't think of anything else right now.

SunPy aims to support a wide range of Python versions:

- SunPy supports Python versions released within the past three years.
- Support for newly released Python versions may take a few months to ensure compatibility.

Deprecation Process
===================

Sunpy's deprecation policy ensures users have ample time to adapt to changes and removals in the API. Below are the key practices:

Deprecation Policy and Breaking Changes
---------------------------------------

Sunpy enforces strict standards for updating, changing, or deprecating public APIs across the core package and stable affiliated packages. If functionality is planned for deprecation, the following steps are taken:

> 1. Use the `@deprecated` decorator to mark code for deprecation.

Reviewer: I don't see the usefulness of exampling how to use the decorator.

Reviewer: Only sections 3 and 5 are needed here.

                           "to maintain the old behaviour or response_format='table' "
                           "to use the new behaviour.")

3. A deprecation warning must be introduced in one LTS release before the deprecated code is removed. For instance:
   - The warning is introduced in SunPy 6.1.
   - The code can only be removed in SunPy 7.1, after the 7.0 LTS release.

4. Add a **“deprecation”** entry to the changelog when introducing the warning and a **“removal”** entry when the code is removed.

5. Every effort is made to avoid breaking changes in public APIs. In cases where it is unavoidable, SunPy conducts code searches on GitHub or GitLab to assess the potential impact on openly available user code.

Documentation of Changes
------------------------

To help users adapt, all API changes are documented thoroughly:

- Deprecation warnings inform users of planned changes or removals, provided an alternative exists or the functionality is being removed entirely.
- All changes must be documented in the CHANGELOG and user documentation.
- When practical, provide a side-by-side comparison of old and new functionality.

> - Include a **“breaking”** changelog entry with detailed guidance for updating user code.

Reviewer: Parts of this should be included in the previous section, it doesn't need to be separate.

.. _sunpy-topic-guide-deprecation-versioning:

> **********************************************
> Deprecation, Versioning and Release Practices
> **********************************************

Reviewer: ```suggestion
*********************************************
Deprecation, Versioning and Release Practices
*********************************************
```

Versioning Policy
=================

Sunpy uses semantic versioning system with three components, each serving a specific purpose:

- **X (LTS Version Number)**: Incremented with every Long Term Support (LTS) release.
- **Y (Release Counter)**: Starts at 0 for LTS releases and increases for intermediate releases.
- **Z (Bug Fix Counter)**: Incremented for bug fix releases.

Here's an example release sequence to illustrate:

- **1.0.0**: An LTS release (November 2024).
- **1.0.1**: A bug fix for the LTS release (December 2024).
- **1.1.0**: A short-support release (May 2025).
- **1.0.2**: A bug fix for the LTS release (June 2025).
- **1.1.1**: A bug fix for the short-support release (June 2025).
- **2.0.0**: An LTS release (November 2025).

> For detailed information, refer to `Sunpy Enhancement Proposal (SEP) 0009 <https://github.com/sunpy/sunpy-SEP/blob/main/SEP-0009.md#deprecations-and-documentation>`__.

Reviewer: ```suggestion
For detailed information, refer to `SunPy Enhancement Proposal (SEP) 0009 <https://github.com/sunpy/sunpy-SEP/blob/main/SEP-0009.md#deprecations-and-documentation>`__.
```

Release Practices
=================

> There will be two planned major releases of ``sunpy`` per year, with approximately 6 months between these releases. Users can expect these releases to occur in the months of June and December, based on the current release schedule of ``astropy``.

Reviewer: ```suggestion
There will be two planned major releases of ``sunpy`` per year, with approximately 6 months between these releases.
Users can expect these releases to occur in the months of June and December, based on the current release schedule of ``astropy``.
```

``sunpy`` provides two types of releases:

- **Long Term Support (LTS) Releases**: These are the first major release of each year and are supported for 12 months or until the next LTS release.
- **Short-Support (Non-LTS) Releases**: These are the second release of the year, supported for 6 months or until the next release.

Support periods may be extended beyond these requirements if necessary.

> Python and Package Support Policies

Reviewer: This is missing the sep-0 note from the other page.

===================================

> Sunpy aims to support a wide range of Python versions:

Reviewer: ```suggestion
``sunpy`` aims to support a wide range of Python versions:
```

> - Sunpy supports Python versions released within the past three years.
> - Support for newly released Python versions may take a few months to ensure compatibility.

Reviewer: ```suggestion
- Supports all Python versions released within the past three years.
- Support for newly released Python versions may take a few months to ensure compatibility but the aim is less than 6 months.
```

> Sunpy has a short list of core dependencies (Python, numpy, astropy, parfive) and a long list of optional dependencies. The minimum version of these packages that we enforce follows this policy:

Reviewer: ```suggestion
``sunpy`` has a short list of core dependencies (Python, numpy, astropy, parfive) and a long list of optional dependencies.
The minimum version of these packages that we enforce follows this policy:
```

- **Python**: Released in the prior 36 months from the anticipated release date.
- **astropy**: Released in the prior 12 months from the anticipated release date.
- **Everything else**: Released in the prior 24 months from the anticipated release date.

> Affiliated packages maintained by the Sunpy team will follow this policy and will support at least the Sunpy LTS version at the time of their release.

Reviewer: ```suggestion
Affiliated packages maintained by the SunPy Project will follow this policy and will support at least the current ``sunpy`` LTS version at the time of their release.
```

> For dependencies only needed to run tests, Sunpy will support versions released in the prior 12 months to the current date, unless there is a critical issue that requires a newer version.

Reviewer: ```suggestion
For dependencies only needed to run tests, ``sunpy`` will support versions released in the prior 12 months to the current date, unless there is a critical issue that requires a newer version.
```

> Deprecation Process
> ===================

Reviewer: ```suggestion
Deprecation Policy and Breaking Changes
=======================================
```

> Sunpy's deprecation policy ensures users have ample time to adapt to changes and removals in the API. Below are the key practices:

Reviewer: ```suggestion
The goal of the deprecation policy is to try and ensure that users have enough time to adapt to any breaking changes or removals in the public API.
Below are the key practices:
```

> Deprecation Policy and Breaking Changes
> ---------------------------------------
> 
> Sunpy enforces strict standards for updating, changing, or deprecating public APIs across the core package and stable affiliated packages. If functionality is planned for deprecation, the following steps are taken:
> 
> - A deprecation warning must be introduced in one LTS release before the deprecated code is removed. For instance:
>   - The warning is introduced in Sunpy 6.1.
>   - The code can only be removed in Sunpy 7.1, after the 7.0 LTS release.
> 
> - Deprecation warnings inform users of planned changes or removals, provided an alternative exists or the functionality is being removed entirely.
> - Include a **“breaking”** changelog entry with detailed guidance for updating user code.
> 
> Every effort is made to avoid breaking changes in public APIs. In cases where it is unavoidable, Sunpy conducts code searches on GitHub or GitLab to assess the potential impact on openly available user code.
> 
> When practical, provide a side-by-side comparison of old and new functionality.

Reviewer: ```suggestion
The SunPy Project enforces these standards for updating, changing, or deprecating public APIs across the core package and **stable** affiliated packages.
If functionality is planned for deprecation, the following steps are taken:

- A deprecation warning must be introduced in one LTS release before the deprecated code is removed.
  For example:
  
    - A deprecation warning is introduced in ``sunpy`` 6.1.
    - The code that is being removed, can only be removed in ``sunpy`` 7.1, after the 7.0 LTS release.

The goal of the deprecation warning is to inform users of breaking changes or removals, provided an alternative exists or the functionality is being removed entirely.

Every effort is made to avoid breaking changes in public APIs.
In cases where it is unavoidable, developers will conduct code searches on GitHub or GitLab to assess the potential impact on openly available user code.

When practical, a migration guide or a side-by-side comparison of old and new functionality will be added to the documentation to ease the transition process.
```

- **Everything else**: Released in the prior 24 months from the anticipated release date.

Affiliated packages maintained by the SunPy Project will follow this policy and will support at least the current ``sunpy`` LTS version at the time of their release.

For dependencies only needed to run tests, ``sunpy`` will support versions released in the prior 12 months to the current date, unless there is a critical issue that requires a newer version.

This policy is based on `SPEC-0 <https://scientific-python.org/specs/spec-0000/>`__.

Deprecation Policy and Breaking Changes
=======================================

The goal of the deprecation warning is to inform users of breaking changes or removals, provided an alternative exists or if the functionality is being removed entirely.

The SunPy Project enforces these standards for updating, changing, or deprecating public APIs across the core package and **stable** affiliated packages.
If functionality is planned for deprecation, the following steps are taken:

A deprecation warning must be introduced in one LTS release before the deprecated code is removed.
For example:
    - A deprecation warning is introduced in ``sunpy`` 6.1.

>     - The code that is being removed, can only be removed in ``sunpy`` 7.1, after the 7.0 LTS release.

Reviewer 2: This introduced a real doc fail, that probably got missed because of aiapy docs being broken: https://github.com/sunpy/sunpy/actions/runs/12678157422/job/35335362996#step:10:450

Author: Hello, sorry for the oversight, I have addressed this a new PR, I hope the checks will pass now!

Reviewer 2: No worries! Thanks for the PR fixing it.

.. _sunpy-topic-guide-deprecation-versioning:

*********************************************
Deprecation, Versioning and Release Practices
*********************************************

Versioning Policy
=================

> Sunpy uses semantic versioning system with three components, each serving a specific purpose:

Reviewer: ```suggestion
``sunpy`` uses a pseudo-semantic versioning system with three components, each serving a specific purpose:
```

- **1.1.0**: A short-support release (May 2025).
- **1.0.2**: A bug fix for the LTS release (June 2025).
- **1.1.1**: A bug fix for the short-support release (June 2025).
- **2.0.0**: An LTS release (November 2025).

For detailed information, refer to `SunPy Enhancement Proposal (SEP) 0009 <https://github.com/sunpy/sunpy-SEP/blob/main/SEP-0009.md#deprecations-and-documentation>`__.

Release Practices
=================

There will be two planned major releases of ``sunpy`` per year, with approximately 6 months between these releases.
Users can expect these releases to occur in the months of June and December, based on the current release schedule of ``astropy``.

``sunpy`` provides two types of releases:

- **Long Term Support (LTS) Releases**: These are the first major release of each year and are supported for 12 months or until the next LTS release.
- **Short-Support (Non-LTS) Releases**: These are the second release of the year, supported for 6 months or until the next release.

Support periods may be extended beyond these requirements if necessary.

Python and Package Support Policies
===================================

> ``sunpy`` aims to support a wide range of Python versions:

Reviewer: ```suggestion
``sunpy`` aims to support:
```

> - Supports all Python versions released within the past three years.

Reviewer: ```suggestion
- All Python versions released within the past three years.
```

> - Support for newly released Python versions may take a few months to ensure compatibility but the aim is less than 6 months.

Reviewer: ```suggestion
- Newly released Python versions within 3 months but may longer due to upstream dependencies.
```

> ``sunpy`` has a short list of core dependencies (Python, numpy, astropy, parfive) and a long list of optional dependencies.

Reviewer: ```suggestion
``sunpy`` has a short list of core dependencies (numpy, astropy, parfive) and a long list of optional dependencies.
```

The minimum version of these packages that we enforce follows this policy:

> - **Python**: Released in the prior 36 months from the anticipated release date.

Reviewer: ```suggestion
```

- **astropy**: Released in the prior 12 months from the anticipated release date.
- **Everything else**: Released in the prior 24 months from the anticipated release date.

Affiliated packages maintained by the SunPy Project will follow this policy and will support at least the current ``sunpy`` LTS version at the time of their release.

For dependencies only needed to run tests, ``sunpy`` will support versions released in the prior 12 months to the current date, unless there is a critical issue that requires a newer version.

> This policy is based on `SPEC-0`_.
> 
> .. _SPEC-0: https://scientific-python.org/specs/spec-0000/

Reviewer: ```suggestion
This policy is based on `SPEC-0 <https://scientific-python.org/specs/spec-0000/>`__.
```

Deprecation Policy and Breaking Changes
=======================================

The goal of the deprecation policy is to try and ensure that users have enough time to adapt to any breaking changes or removals in the public API.

> Below are the key practices:

Reviewer: ```suggestion
```

The SunPy Project enforces these standards for updating, changing, or deprecating public APIs across the core package and **stable** affiliated packages.
If functionality is planned for deprecation, the following steps are taken:

> - A deprecation warning must be introduced in one LTS release before the deprecated code is removed.
>   For example:
> 
>     - A deprecation warning is introduced in ``sunpy`` 6.1.
>     - The code that is being removed, can only be removed in ``sunpy`` 7.1, after the 7.0 LTS release.

Reviewer: ```suggestion

A deprecation warning must be introduced in one LTS release before the deprecated code is removed.
For example:

    - A deprecation warning is introduced in ``sunpy`` 6.1.
    - The code that is being removed, can only be removed in ``sunpy`` 7.1, after the 7.0 LTS release.
```

> The goal of the deprecation warning is to inform users of breaking changes or removals, provided an alternative exists or the functionality is being removed entirely.

Reviewer: ```suggestion
The goal of the deprecation warning is to inform users of breaking changes or removals, provided an alternative exists or if the functionality is being removed entirely.
```
